src/store/Controller.cc

```diff
@@ -837,22 +809,41 @@ Store::Controller::anchorToCache(StoreEntry &entry, bool &inSync)
 
     debugs(20, 7, "anchoring " << entry);
 
+    Transients::EntryStatus entryStatus;
+    transients->status(entry, entryStatus);
+
+    inSync = false;
     bool found = false;
     if (sharedMemStore)
         found = sharedMemStore->anchorToCache(entry, inSync);
     if (!found && swapDir)
         found = swapDir->anchorToCache(entry, inSync);
 
+    if (inSync) {
+        debugs(20, 7, "anchored " << entry);
+        assert(found);
+        entry.setCollapsingRequirement(false);
+        return true;
+    }
+
     if (found) {
-        if (inSync)
-            debugs(20, 7, "anchored " << entry);
-        else
-            debugs(20, 5, "failed to anchor " << entry);
-    } else {
-        debugs(20, 7, "skipping not yet cached " << entry);
+        debugs(20, 5, "failed to anchor " << entry);
+        return true;
+    }
+
+    if (entryStatus.waitingToBeFreed) {
+        debugs(20, 5, "failed on marked unattached " << entry);
+        return true;
     }
 
-    return found;
+    if (!entryStatus.hasWriter) {
+        debugs(20, 5, "failed on abandoned-by-writer " << entry);
+        return true;
```

Comment:
> if entryStatus.hasWriter is false the method always returns after ''if (found)`' condition.

I do not understand why you think that this method returns after `if (found)` condition if entryStatus.hasWriter is false. The two events/states (found and entryStatus.hasWriter) are independent.

Perhaps there is some confusion (on yours and/or my side) because entryStatus is calculated twice (and may change across these two methods)?

Reply:
@eduard-bagdasaryan, did my answer address your concern?

Reply:
> I do not understand why you think that this method returns after if (found) condition if entryStatus.hasWriter is false

IIRC, I considered the following scenario:

* this method (called from Controller::syncCollapsed() ) returns true if entryStatus.hasWriter is false, printing "failed on abandoned-by-writer" in Controller::syncCollapsed()
* a local 'found'  in Controller::syncCollapsed() becomes true, and it proceeds to "aborting unsyncable" (not to "aborting abandoned-by-writer ", which is the next 'if' block).

I am not sure whether this is a problem, i.e., whether these two states "failed on abandoned-by-writer" and "aborting unsyncable " are coherent. On the other hand, we probably should not print the two almost identical messages "failed on abandoned-by-writer" and "aborting abandoned-by-writer ".